### DISCUSSION TITLE
Welcome to DDSP Discussions!

### DISCUSSION THREAD

**Discussion by jesseengel** [magenta/ddsp#209] [Chat / Connect]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

We're trying this out as an experiment. I've enabled the new "Discussions" feature of Github for this repo. Let's try it out as a place to share all things DDSP, connect with each other, share what we've been working on, and discuss. I know these things can get out of hand, so let's give it a trial run and see how it goes.

Try getting started by introducing yourself in *Connect*.

**Comment by zqevans** [comment]
Hey, all! 

I'm Zach, I'm a professional software engineer and a hobbyist music producer. My programming work has largely been web development, and my music is largely electronic. I've tried to learn deep learning a few times before, but once I learned about DDSP, my curiosity really took off. I've been looking for ways to combine music and code, and I've been enamored with the Magenta project since I learned about it last month. Since then, I've been spending as much of my free time as possible learning about ML with a focus on deep learning to enable myself to make use of DDSP. I have a goal of one day writing ML-driven VST plugins, and this seems like a great place to start.

Because I'm learning all of this on my own, I've been looking for people to talk to, work with, and ask questions about deep learning-driven synthesis techniques. I haven't really been able to find an online community focused on ML and audio with respect to music (as opposed to speech recognition), so I'm hoping we can have something like that here.

I'm excited to learn with all of you!

**Reply by alifzp** [reply]
![image](https://user-images.githubusercontent.com/76755211/103325435-16d0d500-4a61-11eb-9964-ccb8c1045f09.png)

can u help me?

**Comment by jesseengel** [comment]
Hey everyone,

I'm Jesse, Magenta research lead, creator of DDSP, and core maintainer of the repo here. I've been working on Magenta for about 4 years now, and besides doing ML, I spend a lot of my time playing guitar (jazz, rock, improv).

Before doing music + ML full time I used to dabble with music tech as a hobby. For example, here's a sound-on-sound looper plugin I made for Ableton Live that also does time-stretching and pitch shifting.

I'll try to triage issues promptly, but apologies if it takes a while. This code repo is also used actively for research internally and a growing list of applications (for instance, g.co/tonetransfer) so it take a bit of work to keep all the parts supported without blocking research. Thanks for your patience :).

**Reply by alifzp** [reply]
hi
i am a student new to everything 
can u please help me about this project?
i have problems even for installing the ddsp library.
ple send your reply to my email.
alifarzipour76@gmail.com

**Reply by osbert-nosa** [reply]
Please who has an acoustic guitar model

**Comment by ethman** [comment]
Hello! 👋👋👋

I'm Ethan. I'm a PhD Candidate in the Interactive Audio Lab at Northwestern University, where I'm happily supervised by Prof. Bryan Pardo. I'm also lucky enough to have spent the past 6 months at Google Brain on the Magenta Team as an intern & student researcher.

Like Jesse, I also play guitar. I used to play a lot of jazz, but these days I've been playing a lot of solo acoustic guitar. I also dabble with bass and piano, though my piano chops are pretty weak these days. 

I've been helping with odds and ends that come up in the DDSP repo and working on my own research on our internal branch (open source version: TBA).

**Reply by alifzp** [reply]
hi
i am a student new to everything
can u please help me about this project?
i have problems even for installing the ddsp library.
ple send your reply to my email.
alifarzipour76@gmail.com

**Reply by ethman** [reply]
Hey, I just saw this. Sorry! If you have a specific question, feel free to post a new discussion question [here](https://github.com/magenta/ddsp/discussions/new) and we'll try to help you out.

**Comment by Forevian** [comment]
I'm Andras, I've spent a decade recording, mixing music in sound studios, then another as software developer. On 2016 dec 24->25 xmas I had a very strange vivid dream about me using a particular plugin in a DAW where the input was midi and out came a realistic flute melody generated by AI. This dream was so powerful that I have enrolled to Andrew Ng's ML Course the very same day. I am currently working for XLN Audio, trying my best to set up a functional ML team.
I've been following audio related research for quite some time now, but the DDSP paper was the first when I felt that my dream might actually come true soon! I enjoy experimenting with DDSP applied to less harmonic inputs, percussive sounds and also for high resolution samples (44 kHz sample rate or even more).
[Hobby music](https://open.spotify.com/track/0JklAnfYMpeutu3wTuLJfX)

**Reply by alifzp** [reply]
hi
i am a student new to everything
can u please help me about this project?
i have problems even for installing the ddsp library.
ple send your reply to my email.
alifarzipour76@gmail.com

**Comment by alifzp** [comment]
guys really need to run this code please help one of you

**Comment by LiftPizzas** [comment]
I'd like to train some more "instruments" for this, and it is implied in several places that this is easy to do, but there doesn't seem to be any simple "here's how to do it yourself" anywhere. Or at least anything that is aimed toward people who are new to this particular programming environment. (Whatever environment it even is.) I don't even know where to begin.

The fact that this has been out for as long as it has and not a single new instrument has been produced, to me suggests I'm not the only one who can't figure it out. Any help would be great.

**Reply by ethman** [reply]
Hello! 

Thanks for checking out the project! Training instructions to make a new instrument are in the README of the [training directory](https://github.com/magenta/ddsp/tree/master/ddsp/training#ddsp-training). Before you train a model, we recommend you have 15-30 minutes of audio of a monophonic instrument (i.e., only one note played at a time) as wav or mp3.

After you [pip install](https://github.com/magenta/ddsp/blob/master/README.md#installation) the prerequisites, the instructions to train a new instrument are [here](https://github.com/magenta/ddsp/tree/master/ddsp/training#training-a-model-on-your-own-data). To train a new instrument model, you will need access to a GPU or TPU. If you don't have one at the ready, AWS offers EC2 instances with GPU access and Google Compute Engine also offers a "Deep Learning VM" that is ready made for making a new instrument with DDSP. Both of these options could potentially cost money, however.

As the networks train, the Spectral Loss will get output to the console; we have experienced that the instruments sound best when the spectral loss is around 4-5. We expect that this will take about 12-24 hours, depending on the hardware. Checkpoints are stored periodically as the networks train. Once your network is trained, you can upload the checkpoints into [this Colab notebook](https://colab.research.google.com/github/magenta/ddsp/blob/master/ddsp/colab/demos/timbre_transfer.ipynb#scrollTo=wmSGDWM5yyjm) to play around with it.

If you want to dive deep into the library and how DDSP works, we have provided a set of tutorials [here](https://github.com/magenta/ddsp/tree/master/ddsp/colab/tutorials) (these tutorials are [Google Colab notebooks](https://colab.research.google.com/notebooks/intro.ipynb)). While you technically can train in a Colab notebook, they don't provide the best experience (on the free version, you get booted off your GPU after a certain amount of time). 

Finally, we released a set of pre-trained models as part of the [Tone Transfer](https://sites.research.google/tonetransfer) project if you want to mess around in the mean time.

I hope that helps. Feel free to reach out if you have any other questions.

-Ethan

**Reply by LiftPizzas** [reply]
Hi Ethan, thanks for your response. :) The Tone Transfer project is what led me here. I like what is there and it's a cool tech demo, but it's also extremely narrow in scope compared to what it could be. The possibilities are huge and to only have a couple of instruments is a tragedy. I wanted to expand the number of possible instruments usable as output, and I have the equipment needed to do so.

All of the documents are devoid of context and assume I'm already in the same environment. I can't tell which OS, and/or which development environment any of these things are supposed to be happening in. I can't tell but it looks like there are several prerequisite pieces of software that would already need to be installed before you even get to the first readme.

I'm on a windows 10 machine with a geforce 2070 GPU. I have various development environments including MS Visual Studio, Android Studio, Unity, Unreal engine 4, and Eclipse. I also have an old PC/GPU I can throw another OS on. I'm sure I can get whatever it is working if I have instructions. We also have several instruments and recording equipment. Producing the input wouldn't be a problem and I'm fine with (and would actually enjoy) plenty of trial and error to see the output coming from various inputs.

I'm not so interested in a deep dive of DDSP or ANNs at this point. My difficulty here is in figuring out how to get a PC set up to run whatever software is needed to do the training.

**Reply by Forevian** [reply]
2070 is nice, consider setting up a Linux dual boot (ie. Ubuntu) on your machine if you want to use it for training neural networks, and make sure it has CUDA installed. It might be quite techy (and time consuming) to get to the state where you can run your experiments utilizing your GPU, but once it works, you can have all the fun. Unfortunately, it is still easier to go this way compared to setting up everything to work properly under windows. Using the cloud for training is actually the easiest, and yes, it costs money, but at least then you don't need to worry about the noise/heat produced by your machine and can use it for other stuff while training is happening. I hope I didn't misunderstand your situation, if so, just ignore me :)

**Reply by ethman** [reply]
Yeah, like Forevian implied, most neural network models are trained on Ubuntu these days. But it's certainly possible to use your Windows machine, as long as your GPU supports CUDA. Getting your machine setup with all of the drivers with CUDA can be challenging, which is why both Forevian and I recommended using a cloud machine. That being said, it's not impossible. It might be the case that you'll have to dedicate a half day to googling driver installation errors to try to debug, but this only needs to be done once. Once it's set up, you're good to go. Here are [the official TF instructions](https://www.tensorflow.org/install/gpu) for getting your machine set up. I'm also assuming you're all good with python set up as well, but this has historically been tricky to set up on Windows (though might not be as much anymore). If you need to set up python, I recommend using [Anaconda](https://www.anaconda.com/products/individual#Downloads).

All this being said, we are aware of how much of a challenge this whole process is. The path from audio --> new model is not so straight forward. We are working on making this process easier in future releases. Stay tuned!

**Reply by LiftPizzas** [reply]
Thanks, I wouldn't have guessed linux or python. I haven't worked with either. (Well I installed linux back in 2002 but never found a use for it at the time.) I've got an older machine with a blank HDD and a GTX 960 that I can turn into a linux box, then at least it won't be tying up my desktop.

**Reply by LiftPizzas** [reply]
I installed Ubuntu on a box with a GTX960. ... 

Edited to update: I seem to have everything but cuda installed. Based on the output I got, I'm not sure if I will know if I got everything installed or if it's working like it should. Is there by chance one of the samples of training data to use as input?

**Comment by harveyfretwell** [comment]
Hey,

I'm currently studying Sound Engineering and Production at BCU.

I was looking at using DDSP in my final year project but wanted to check if what I'm trying to do would even be possible first.
I wanted to use DDSP to perform automatic phase alignment on an audio signal using a reference signal to align it with.

Would this be possible?
